contracts/modules/cover/Cover.sol

```diff
@@ -0,0 +1,775 @@
+    require(productIds.length == initialPriceRatios.length, "Cover: Array lengths must not be different");
+    for (uint i = 0; i < productIds.length; i++) {
+      require(initialPriceRatios[i] >= GLOBAL_MIN_PRICE_RATIO, "Cover: Initial price must be greater than the global min price");
+      _products[productIds[i]].initialPriceRatio = initialPriceRatios[i];
+    }
+  }
+
+  function setCapacityReductionRatio(uint productId, uint16 reduction) external onlyAdvisoryBoard {
+    require(reduction <= CAPACITY_REDUCTION_DENOMINATOR, "Cover: LTADeduction must be less than or equal to 100%");
+    _products[productId].capacityReductionRatio = reduction;
+  }
+
+  function addProducts(
+    Product[] calldata newProducts,
+    string[] calldata ipfsMetadata
+  ) external override onlyAdvisoryBoard {
+    uint initialProductsCount = _products.length;
+    for (uint i = 0; i < newProducts.length; i++) {
+      _products.push(newProducts[i]);
+      emit ProductUpserted(initialProductsCount + i, ipfsMetadata[i]);
+    }
+  }
+
+  function editProductsIpfsMetadata(
+    uint[] calldata productIds,
+    string[] calldata ipfsMetadata
+  ) external override onlyAdvisoryBoard {
+    for (uint i = 0; i < productIds.length; i++) {
+      emit ProductUpserted(productIds[i], ipfsMetadata[i]);
+    }
+  }
+
+  function addProductTypes(
+    ProductType[] calldata newProductTypes,
+    string[] calldata ipfsMetadata
+  ) external override onlyAdvisoryBoard {
+    uint initialProuctTypesCount = _productTypes.length;
+    for (uint i = 0; i < newProductTypes.length; i++) {
+      _productTypes.push(newProductTypes[i]);
+      emit ProductTypeUpserted(initialProuctTypesCount + i, ipfsMetadata[i]);
+    }
+  }
+
+
+  /* ========== ACTIVE COVER AMOUNT TRACKING ========== */
+
+  function enableActiveCoverAmountTracking(uint24[] memory assetIds, uint[] memory activeCoverAmountsForAssets) external onlyEmergencyAdmin {
+    require(!activeCoverAmountCommitted, "Cover: activeCoverAmountCommitted = true");
+
+    require(assetIds.length == activeCoverAmountsForAssets.length, "Cover: Array lengths must not be different");
+    if (!coverAmountTrackingEnabled) {
+      coverAmountTrackingEnabled = true;
+    }
+
+
+    for (uint i = 0; i < assetIds.length; i++) {
+      totalActiveCoverInAsset[assetIds[i]] = activeCoverAmountsForAssets[i];
+    }
+  }
+
+  function commitActiveCoverAmounts() external onlyEmergencyAdmin {
+    require(!activeCoverAmountCommitted, "Cover: activeCoverAmountCommitted = true");
+    activeCoverAmountCommitted = true;
+  }
+
+  /* ========== HELPERS ========== */
+
+  function isAssetSupported(uint32 payoutAssetsBitMap, uint8 payoutAsset) public view override returns (bool) {
+
+    if (payoutAssetsBitMap == 0) {
+      return (1 << payoutAsset) & coverAssetsFallback > 0;
+    }
+    return (1 << payoutAsset) & payoutAssetsBitMap > 0;
+  }
+
+  /* ========== CAPACITY CALCULATION ========== */
+
+  function calculateCapacity(
```

Comment:
seems to only be used in the viewer contract. remove from here?

Reply:
done